executor/sample.go

```diff
@@ -0,0 +1,365 @@
+	schema *expression.Schema, fullSchema *expression.Schema, retTypes []*types.FieldType, desc bool) *tableRegionSampler {
+	return &tableRegionSampler{
+		ctx:        ctx,
+		table:      t,
+		startTS:    startTs,
+		partTables: partTables,
+		schema:     schema,
+		fullSchema: fullSchema,
+		isDesc:     desc,
+		retTypes:   retTypes,
+		rowMap:     make(map[int64]types.Datum),
+	}
+}
+
+func (s *tableRegionSampler) writeChunk(req *chunk.Chunk) error {
+	if s.restKVRanges == nil {
+		var err error
+		s.restKVRanges, err = s.splitTableRanges()
+		if err != nil {
+			return err
+		}
+		sortRanges(s.restKVRanges, s.isDesc)
+	}
+	var regionKeyRanges []kv.KeyRange
+	cutPoint := mathutil.MinInt64(int64(req.RequiredRows()), int64(len(s.restKVRanges)))
+	regionKeyRanges, s.restKVRanges = s.restKVRanges[:cutPoint], s.restKVRanges[cutPoint:]
+
+	decLoc, sysLoc := s.ctx.GetSessionVars().TimeZone, time.UTC
+	cols, decColMap, err := s.buildSampleColAndDecodeColMap()
+	if err != nil {
+		return err
+	}
+	rowDecoder := decoder.NewRowDecoder(s.table, cols, decColMap)
+	err = s.scanFirstKVForEachRange(regionKeyRanges, func(handle kv.Handle, value []byte) error {
+		_, err := rowDecoder.DecodeAndEvalRowWithMap(s.ctx, handle, value, decLoc, sysLoc, s.rowMap)
+		if err != nil {
+			return err
+		}
+		currentRow := rowDecoder.CurrentRowWithDefaultVal()
+		mutRow := chunk.MutRowFromTypes(s.retTypes)
+		for i, col := range s.schema.Columns {
+			offset := decColMap[col.ID].Col.Offset
+			target := currentRow.GetDatum(offset, s.retTypes[i])
+			mutRow.SetDatum(i, target)
+		}
+		req.AppendRow(mutRow.ToRow())
+		s.resetRowMap()
+		return nil
+	})
+	if len(s.restKVRanges) == 0 {
+		s.isFinished = true
+	}
+	return err
+}
+
+func (s *tableRegionSampler) splitTableRanges() ([]kv.KeyRange, error) {
+	if len(s.partTables) != 0 {
+		var ranges []kv.KeyRange
+		for _, t := range s.partTables {
+			for _, pid := range t.GetAllPartitionIDs() {
+				start := tablecodec.GenTableRecordPrefix(pid)
+				end := start.PrefixNext()
+				rs, err := splitIntoMultiRanges(s.ctx.GetStore(), start, end)
+				if err != nil {
+					return nil, err
+				}
+				ranges = append(ranges, rs...)
+			}
+		}
+		return ranges, nil
+	}
+	startKey, endKey := s.table.RecordPrefix(), s.table.RecordPrefix().PrefixNext()
+	return splitIntoMultiRanges(s.ctx.GetStore(), startKey, endKey)
+}
+
+func splitIntoMultiRanges(store kv.Storage, startKey, endKey kv.Key) ([]kv.KeyRange, error) {
+	kvRange := kv.KeyRange{StartKey: startKey, EndKey: endKey}
+
+	s, ok := store.(tikv.Storage)
+	if !ok {
+		return []kv.KeyRange{kvRange}, nil
+	}
+
+	maxSleep := 10000 // ms
+	bo := tikv.NewBackofferWithVars(context.Background(), maxSleep, nil)
+	var ranges []kv.KeyRange
+	regions, err := s.GetRegionCache().LoadRegionsInKeyRange(bo, startKey, endKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, r := range regions {
+		ranges = append(ranges, kv.KeyRange{StartKey: r.StartKey(), EndKey: r.EndKey()})
```

Comment:
Is it possible that `r.StartKey()` is less than `startKey`, `r.EndKey()` is greater than `endKey`?